GTG/core/tasks2.py

```diff
@@ -0,0 +1,559 @@
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program.  If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+"""Everything related to tasks."""
+
+
+from gi.repository import GObject
+from gettext import gettext as _
+
+from uuid import uuid4, UUID
+import logging
+from typing import Callable, Any, Optional
+from enum import Enum
+import re
+import datetime
+from operator import attrgetter
+
+from lxml.etree import Element, SubElement, CDATA
+
+from GTG.core.base_store import BaseStore
+from GTG.core.tags2 import Tag2, TagStore
+from GTG.core.dates import Date
+
+log = logging.getLogger(__name__)
+
+
+# ------------------------------------------------------------------------------
+# REGEXES
+# ------------------------------------------------------------------------------
+
+TAG_REGEX = re.compile(r'^\B\@\w+(\-\w+)*\,+')
+SUB_REGEX = re.compile(r'\{\!.+\!\}')
+
+
+# ------------------------------------------------------------------------------
+# TASK STATUS
+# ------------------------------------------------------------------------------
+
+class Status(Enum):
+    """Status for a task."""
+
+    ACTIVE = 'Active'
+    DONE = 'Done'
+    DISMISSED = 'Dismissed'
+
+
+class Filter(Enum):
+    """Types of filters."""
+
+    STATUS = 'Status'
+    TAG = 'Tag'
+    PARENT = 'Parent'
+    CHILDREN = 'Children'
+
+
+# ------------------------------------------------------------------------------
+# TASK
+# ------------------------------------------------------------------------------
+
+class Task2(GObject.Object):
+    """A single task."""
+
+    __gtype_name__ = 'Task'
+    __slots__ = ['id', 'raw_title', 'content', 'tags',
+                 'children', 'status', 'parent', '_date_added',
+                 '_date_due', '_date_start', '_date_closed',
+                 '_date_modified']
+
+
+    def __init__(self, id: UUID, title: str) -> None:
+        self.id = id
+        self.raw_title = title.strip('\t\n')
+        self.content =  ''
+        self.tags = []
+        self.children = []
+        self.status = Status.ACTIVE
+        self.parent = None
+
+        self._date_added = Date.no_date()
+        self._date_due = Date.no_date()
+        self._date_start = Date.no_date()
+        self._date_closed = Date.no_date()
+        self._date_modified = Date(datetime.datetime.now())
+
+
+    def toggle_status(self, propagate: bool = True) -> None:
+        """Toggle between possible statuses."""
```

Comment:
I think we should have separate `set_status` so you can try to close an already closed task and it does nothing. Like `dismiss()`.

Reply:
Yeah, that might be useful for completions sake